### DISCUSSION TITLE
Rollup through junction object

### DISCUSSION THREAD

**Discussion by mhogg-srg** [jamessimone/apex-rollup#282] [General]
I have a situation where I'm using a junction object to create a many-to-many relationship between two objects. Let's call the main objects A and B and the junction J. I'd like to take a values from A and roll them into B. Is it possible to do this without first creating formula fields on J that output values directly from A?

**Comment by jamessimone** [comment]
@mhogg-srg converted this to a discussion, though I may re-create the issue should the need arise. Let's talk about the situation here, and what you're looking to do.

Let's say you have 5 records for Object A, with 5 junction records in J. They're all linked to one record in Object B. That's the situation we're referring to, yes? If I'm understanding you correctly, you'd like to be able to roll up the values from A -> J -> B. That _should_ be possible using a grandparent rollup, though I have not tested this with a many-to-many junction previously. Have you checked out [the wiki article for this](https://github.com/jamessimone/apex-rollup/wiki/Configuring-Grandparent-(or-Greater)-Rollups)?

Again - it's possible there are some edge-cases not considered in the existing grandparent rollup framework, and I'm happy to get an issue going for those after chatting with you here, but let's first try out using the grandparent rollups. If it doesn't work, it would be great to enable `Is Rollup Logging Enabled` on your Rollup Control record and get the Apex Debug Logs generated from attempting this attached here.

Looking forward to your response!

**Reply by mhogg-srg** [reply]
I should've made this a discussion to begin with. Sorry about that.

The scenario I'm working on is not a grandparent relationship. J is a child of both A and B--it has a lookup field for A and another for B. So the type of links from A to B are _child, parent_ rather than _parent, parent_.

**Reply by jamessimone** [reply]
@mhogg-srg sorry for the delay in reply. While it may not be a traditional hierarchy, I’m still of the mind that the grandparent rollup will work for A -> J -> B.

If there are multiple J’s related to A, and those J’s are related to multiple B’s, we might be talking “feature request” to get everything working properly. It’s an interesting problem to think about, so if that is indeed the situation please let me know.

I guess for a “grandparent rollup” (or, in this case, “a rollup with more than one relationship hop to the ultimate parent”) where the intermediate parent is technically a collection of children records, the existing grandparent field path would probably “just work” (without the need for any more configuration or extra fields) if I were to add this functionality, since the “field path” (such as it is) remains the same. Let me know what you think!

**Reply by mhogg-srg** [reply]
I'm thinking that probably my original idea of putting formula fields on J is the simplest and most pragmatic way forward. However, this conversation may be fruitful.

The obstacle I'm running into with your proposed solution is Lookup Field on Calc Item. Since J is a child of A, there's no lookup field on A to select.

**Reply by jamessimone** [reply]
Ah of course. Ok, let me think about this — you may be right that short-term the formula fields are the way to go, but I’ll definitely be able to support your original use case in the near future

**Reply by jamessimone** [reply]
@mhogg-srg I have a prototype version of this working locally. I am going to keep the `Lookup Field On Calc Item` a required field, since it will be for the vast majority of rollup use-cases; for this _specific_ case, you can use any field there since it won't really be looked at. Here's my proposal:

- fill out the `Grandparent Relationship Field Path`
- also fill out a new field, `One To Many Grandparent Fields`

As an example, I've created a test that tests going from Account -> Contacts -> Individual (totally contrived example):

- my grandparent field path is: `Contact.Individual.Name`
- my one to many grandparent field (which can be a comma-separated list if you need to hop through multiple junction objects) contains the object name and the field API name: `Contact.AccountId`

How does that kind of mapping sound to you?

**Reply by mhogg-srg** [reply]
That solution makes a lot of sense. The one thing I find a bit unintuitive is using `Contact.Individual.Name` rather than `Contacts.Individual.Name` (notice the s), which is the actual name of the relationship going from Account -> Contact. Using `Contact.AccountId` in the `One to Many Grandparent Fields` does make sense to me since it's naming the object itself.

**Reply by jamessimone** [reply]
It is unintuitive, I agree - I can update that to use the relationship name. It's a bit more code, but it will be more intuitive as to how the grandparent field path should be filled out. Thanks so much for the feedback! I will create an issue to track this and tag you in it, along with #283.

**Comment by mhogg-srg** [comment]
@jamessimone I can't figure out what I'm doing wrong. When I run Recalculate Rollup with this, the job completes successfully, but none of the values are updated. Field_Leadership__c is the junction object with two parent-detail relationships: Leader__c to Contact, and Field_Sector__c to Team__c. Any recommendations?

<img width="919" alt="Screen Shot 2022-04-18 at 11 10 18 AM" src="https://user-images.githubusercontent.com/63794410/163829294-34b71b19-cec4-414c-aeb6-82a6a2485763.png">

**Reply by jamessimone** [reply]
@mhogg-srg I don't think you're doing anything wrong - I need to check on the Recalculate Rollup app as there seems to be something unexpected happening there. Give me a few hours to address this!

**Reply by jamessimone** [reply]
@mhogg-srg just to confirm - in your grandparent relationship field path, is `Field_Sector__r` a relationship field on your `Field_Leadership__c` junction object?

**Comment by mhogg-srg** [comment]
That's correct.

**Reply by jamessimone** [reply]
@mhogg-srg I'm generating a beta package with some updates, with the intent to release the new version tomorrow. If you're doing any of this testing in a sandbox, you should be able to access the [updated Readme](https://github.com/jamessimone/apex-rollup/tree/v1.4.13) with the beta package link in it within the next few minutes. Note that the above link will cease to function once I've merged the changes to Apex Rollup's `main` branch (which, again, I'm planning to do as part of getting this release out the door tomorrow).

If you _are_ able to test before I come online again, great; otherwise, I'll wait a few hours tomorrow prior to releasing.

**Reply by mhogg-srg** [reply]
@jamessimone It worked!

**Reply by jamessimone** [reply]
@mhogg-srg great! I'm going to go ahead and cut a release shortly.
